garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+            - A child model is built.
+            - A parent model wit the same name exists.
+
+        Returns:
+          outputs: Output tensor of the model with the given inputs.
+
+        """
+        if not name:
+            if 'default' in self._siblings:
+                raise ValueError("Subgraph {} already exists!".format(
+                    self.name))
+            with self._name_scope:
+                if not self._is_built:
+                    _variable_scope = self.variable_scope
+                else:
+                    _variable_scope = tf.variable_scope(
+                        self._variable_scope._name_or_scope, reuse=True)
+                with _variable_scope:
+                    self._inputs = inputs
+                    outputs = self._build(*inputs)
+                    self._outputs = outputs
+                    self._siblings['default'] = self
+                    if not self._is_built:
+                        variables = self._get_variables().values()
+                        tf.get_default_session().run(
+                            tf.variables_initializer(variables))
+                        if self._default_parameters:
+                            self.parameters = self._default_parameters
+                        self._is_built = True
+        elif name in self._siblings:
+            raise ValueError("Subgraph {} already exists!".format(name))
+        elif not self._is_parent:
+            raise ValueError("Child of a model should not be built!")
+        else:
+            twin = type(self)._new_child(self, name)
+            setattr(self, name, twin)
+            self._siblings[name] = twin
+            outputs = twin.build(*inputs)
+            if not self._is_built:
+                self._inputs = inputs
+                self._outputs = list(outputs)
+                self._is_built = True
+
+        return outputs
+
+    def _build(self, *inputs):
+        """
+        Output of the model given input placeholder(s).
+
+        User should implement _build inside their subclassed model.
+        """
+        pass
+
+    @property
+    def variable_scope(self):
+        """Variable scope of the model."""
+        return self._variable_scope
+
+    @property
+    def name_scope(self):
+        """Name scope of the model."""
+        return self._name_scope
+
+    @property
+    def parameters(self):
+        """Parameters of the model."""
+        return tf.get_default_session().run(self._get_variables())
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        variables = self._get_variables()
+        for name, var in variables.items():
+            if name in parameters:
+                var.load(parameters[name])
+            else:
+                warnings.warn('No value provided for variable {}'.format(name))
+
+    def _get_variables(self):
+        return {
+            v.name: v
+            for v in tf.global_variables(
```

Comment:
there is a new way of doing this in 1.12

https://www.tensorflow.org/api_docs/python/tf/VariableScope#get_collection
https://www.tensorflow.org/api_docs/python/tf/VariableScope#global_variables